lib/spreedly/environment.ex

```diff
@@ -1,156 +1,210 @@
 defmodule Spreedly.Environment do
+  @moduledoc """
+  An Elixir client implementation of the Spreedly API.
 
-  defstruct [ :environment_key, :access_secret ]
+  For more info visit the [Spreedly API docs](https://docs.spreedly.com/reference/api/v1/)
+  for a detailed listing of available API methods.
 
-  import Spreedly.URL
-  import Spreedly.RequestBody
+  ## Usage
 
+  API interactions happen on a `Spreedly.Environment`.
+
+      iex> env = Spreedly.Environment.new(environment_key, access_secret)
+
+  Once you have an environment, you can use it to interact with the API.
+
+  ### Run a purchase using a credit card
+
+  You can pattern match on the response.
+
+      iex> case Spreedly.Environment.purchase(env, "R8AKGmYwkZrrj2BpWcPge", "RjTFFZQp4MrH2HJNfPwK", 2344) do
+              {:ok, %{succeeded: true}} ->
+                IO.puts "Success!"
+              {:ok, %{succeeded: false, message: msg}} ->
+                IO.puts "Declined!"
+              {:error, reason} ->
+                IO.inspect reason
+            end
+
+  ### Show a Transaction
+
+      iex> Spreedly.Environment.show_transaction(env, "7f6837d1d22e049f8a47a8cc1fa9")
+      {:ok,
+      %{created_at: "2016-01-10T16:36:14Z", currency_code: nil, description: nil,
+        ...
+        state: "gateway_processing_failed", succeeded: false,
+        token: "7f6837d1d22e049f8a47a8cc1fa9", transaction_type: "Verification",
+        updated_at: "2016-01-10T16:36:14Z"}}
+
+      iex> Spreedly.Environment.find_transaction(env, "NonExistentToken")
+      {:error, "Unable to find the transaction NonExistentToken."}
+  """
+
+  import Spreedly
+  import Spreedly.{Path, RequestBody}
+
+  defstruct [:environment_key, :access_secret]
+
+  @type t :: %__MODULE__{environment_key: String.t, access_secret: String.t}
+
+  @spec new(String.t, String.t) :: t
   def new(environment_key, access_secret) do
-    %Spreedly.Environment{environment_key: environment_key, access_secret: access_secret}
+    %__MODULE__{environment_key: environment_key, access_secret: access_secret}
   end
 
+  @spec add_gateway(t, String.t) :: {:ok, any} | {:error, any}
   def add_gateway(env, gateway_type) do
-    HTTPoison.post(add_gateway_url(), add_gateway_body(gateway_type), headers(env))
-    |> response
+    post_request(env, add_gateway_path(), add_gateway_body(gateway_type))
```

Comment:
I ❤️ these simplifications.

Reply:
I wanted to have `get` and `post`, etc, but they conflicted with HTTPPoison Base (caught by dialyzer).